### PR TITLE
Adds support to test custom payload branches

### PR DESCRIPTION
**Pull request by cgranleese-r7**
This PR adds support to allow us to test payload branches as part of our CI acceptance testing. If we wanted to test payload changes in Framework before they're merged and released as gems separately, that can now be done by adding a the `payload-testing-branch` for metasploit-payloads payloads, and/or `payload-testing-mettle-branch` for Mettle payloads.

## Verification

- [ ] CI goes green
- [ ] Code changes are sane
- [ ] Can point the GitHub actions at custom payload branches (WILL REQUIRE GITHUB LABELS TO BE ADDED FIRST)

### CODE REVIEW

File: .github/workflows/acceptance.yml [hunk 1]
```diff
@@ -96,7 +108,7 @@ jobs:
         if: runner.os == 'Linux'
         run: sudo apt-get -y --no-install-recommends install libpcap-dev graphviz
 
-      - uses: shivammathur/setup-php@fc14643b0a99ee9db10a3c025a33d76544fa3761
+      - uses: shivammathur/setup-php@6d7209f44a25a59e904b1ee9f3b0c33ab2cd888d
```
Comment (adfoster-r7) [inline .github/workflows/acceptance.yml:111]: I believe this can be reverted

File: data/cmd_exec/makefile.mk [hunk 2]
```diff
@@ -0,0 +1,5 @@
+all: show_args_linux show_args_windows
+show_args_linux: show_args.c
+	cc show_args.c -o show_args
```
Comment (adfoster-r7) [inline data/cmd_exec/makefile.mk:3]:
```suggestion
	cc show_args.c -o show_args_linux
```

File: data/cmd_exec/show_args
Comment (adfoster-r7) [inline data/cmd_exec/show_args:1]: `show_args_linux`

File: spec/acceptance/README.md [hunk 3]
```diff
@@ -27,9 +27,14 @@ SPEC_OPTS='--tag acceptance' SPEC_HELPER_LOAD_METASPLOIT=false bundle exec rspec
 ```
 
 Run a specific Meterpreter/module test Unix / Windows:
+
+# Bash command:
```
Comment (adfoster-r7) [inline spec/acceptance/README.md:31]: The headings here isn't hierarchically correctly 👀 

I think either more `#` needs added to fix the header depth, or maybe it shouldn't be a heading as the other examples aren't 🤔

File: data/cmd_exec/show_args
Comment (cgranleese-r7) [reply data/cmd_exec/show_args:1]: Thanks for taking a look 👍 

I'm assuming this was supposed to be part of the [comment](https://github.com/rapid7/metasploit-framework/pull/19390#discussion_r1717487406) above or something 👀

File: test/modules/post/test/cmd_exec.rb [hunk 4]
```diff
@@ -37,6 +57,29 @@ def test_cmd_exec
       output == test_string
     end
 
+    it 'should execute show_args_* executables and return the passed arguments' do
+      if session.platform.eql? 'windows'
```
Comment (adfoster-r7) [inline test/modules/post/test/cmd_exec.rb:61]: Is there a cleaner way to write this? 👀 

It's a bit unweldy to read at a quick glance to understand what it's doing with the nested complexity here

Comment (adfoster-r7) [reply test/modules/post/test/cmd_exec.rb:61]: Here's a potential golf: 
https://github.com/adfoster-r7/metasploit-framework/commit/adc25ccbc941eb299fdedaf1203bb344dea79a88

File: .github/workflows/acceptance.yml [hunk 5]
```diff
@@ -129,8 +141,64 @@ jobs:
...
 
-      - name: Checkout code
+      # The job checkout structure is:
+      #  .
+      #  ├── metasploit-framework
+      #  └── metasploit-payloads (Only if metasploit-payloads-testing-branch is applied)
+      #  └── mettle (Only if mettle-testing-branch is applied)
+
+      - name: Install Docker - macOS
+        if: ${{ ( matrix.meterpreter.name == 'java') && (runner.os == 'macos' ) && (contains(github.event.issue.labels.*.name, 'mettle-testing-branch')) }}
```
Comment (adfoster-r7) [inline .github/workflows/acceptance.yml:151]:
```suggestion
        if: ${{ ( matrix.meterpreter.name == 'java') && (runner.os == 'macos' ) && (contains(github.event.issue.labels.*.name, 'payload-testing-branch')) }}
```

File: .github/workflows/acceptance.yml [hunk 6]
```diff
@@ -81,10 +91,12 @@ jobs:
 
     runs-on: ${{ matrix.os }}
 
-    timeout-minutes: 25
+    timeout-minutes: 50
```
Comment (adfoster-r7) [inline .github/workflows/acceptance.yml:94]: Not a blocker; we could make this conditional - so 25 by default, 50 for custom

File: .github/workflows/acceptance.yml (hunk 5 above)
Comment (adfoster-r7) [inline .github/workflows/acceptance.yml:151]: Can we go with: `payload-testing-mettle-branch` and `payload-testing-branch`

File: .github/workflows/acceptance.yml [hunk 7]
```diff
@@ -129,8 +141,64 @@ jobs:
...
+        run: |
+          echo "METTLE_VERSION=$(grep -oh '[0-9].[0-9].[0-9]*' lib/metasploit_payloads/mettle/version.rb)" >> $GITHUB_ENV
+        working-directory: mettle
+
+      - name: Use mettle version
+        if: ${{ matrix.meterpreter.name == 'mettle' && (contains(github.event.issue.labels.*.name, 'mettle-testing-branch')) }}
+        run: |
+          echo "${{ env.METTLE_VERSION }}"
+        working-directory: mettle
+
```
Comment (adfoster-r7) [inline .github/workflows/acceptance.yml:176]:
```suggestion
```

File: .github/workflows/acceptance.yml [hunk 8]
```diff
@@ -129,8 +141,64 @@ jobs:
...
         uses: actions/checkout@v4
+        with:
+          repository: rapid7/mettle
+          path: mettle
+          ref: ${{ env.mettleCommit }}
+
+      - name: Get mettle version
+        if: ${{ matrix.meterpreter.name == 'mettle' && (contains(github.event.issue.labels.*.name, 'mettle-testing-branch')) }}
+        run: |
+          echo "METTLE_VERSION=$(grep -oh '[0-9].[0-9].[0-9]*' lib/metasploit_payloads/mettle/version.rb)" >> $GITHUB_ENV
```
Comment (adfoster-r7) [inline .github/workflows/acceptance.yml:168]:
```suggestion
          echo "METTLE_VERSION=$(grep -oh '[0-9].[0-9].[0-9]*' lib/metasploit_payloads/mettle/version.rb)" | tee -a $GITHUB_ENV
```

File: .github/workflows/acceptance.yml [hunk 9]
```diff
@@ -139,12 +207,75 @@ jobs:
         with:
           ruby-version: ${{ matrix.ruby }}
           bundler-cache: true
-          cache-version: 4
+          working-directory: metasploit-framework
+          cache-version: 5
```
Comment (adfoster-r7) [inline .github/workflows/acceptance.yml:211]:
```suggestion
          cache-version: 4
          working-directory: metasploit-framework
```

File: .github/workflows/acceptance.yml [hunk 10]
```diff
@@ -139,12 +207,75 @@ jobs:
...
+        if: ${{ matrix.meterpreter.name == 'mettle' && runner.os == 'macos' && (contains(github.event.issue.labels.*.name, 'mettle-testing-branch')) }}
+        run: |
+          cp /Users/runner/work/metasploit-framework/metasploit-framework/mettle/pkg/metasploit_payloads-mettle-${{ env.METTLE_VERSION }}.pre.dev.gem /Users/runner/work/metasploit-framework/metasploit-framework/metasploit-framework
+        working-directory: metasploit-framework
+
+      - name: Move mettle gem into framework - macOS
+        if: ${{ matrix.meterpreter.name == 'mettle' && runner.os != 'macos' && (contains(github.event.issue.labels.*.name, 'mettle-testing-branch')) }}
+        run: |
+          cp /home/runner/work/metasploit-framework/metasploit-framework/mettle/pkg/metasploit_payloads-mettle-${{ env.METTLE_VERSION }}.pre.dev.gem /home/runner/work/metasploit-framework/metasploit-framework/metasploit-framework
+        working-directory: metasploit-framework
```
Comment (adfoster-r7) [inline .github/workflows/acceptance.yml:226]:
```suggestion
      - name: Move mettle gem into framework
        if: ${{ matrix.meterpreter.name == 'mettle' && (contains(github.event.issue.labels.*.name, 'mettle-testing-branch')) }}
        run: |
          cp ./mettle/pkg/metasploit_payloads-mettle-${{ env.METTLE_VERSION }}.pre.dev.gem ./metasploit-framework
```

You might need an extra layer of metasploit-framework here, but I don't think so, i.e. prefixing with an additional `metasploit-framework` to get `./metasploit-framework/mettle`

File: .github/workflows/acceptance.yml [hunk 11]
```diff
@@ -139,12 +207,75 @@ jobs:
...
+          repository: rapid7/metasploit-payloads
+          path: metasploit-payloads
+          ref: ${{ env.metasploitPayloadsCommit }}
+
+      - name: Build Java & Android payloads
+        if: ${{ (matrix.meterpreter.name == 'java') && (runner.os != 'Windows') && (contains(github.event.issue.labels.*.name, 'metasploit-payloads-testing-branch')) }}
+        run: |
+          cd ..
+          docker run --rm -w "$(pwd)" -v "$(pwd):$(pwd)" rapid7/msf-ubuntu-x64-meterpreter:latest /bin/bash -c "cd metasploit-payloads/java && make clean && make android && mvn -P deploy package"
+        working-directory: metasploit-payloads
```
Comment (adfoster-r7) [inline .github/workflows/acceptance.yml:252]:
```suggestion
          docker run --rm -w "$(pwd)" -v "$(pwd):$(pwd)" rapid7/msf-ubuntu-x64-meterpreter:latest /bin/bash -c "cd metasploit-payloads/java && make clean && make android && mvn -P deploy package"
```

File: .github/workflows/acceptance.yml [hunk 12]
```diff
@@ -139,12 +207,75 @@ jobs:
...
+
+      - name: Checkout metasploit-payloads
+        if: contains(github.event.issue.labels.*.name, 'metasploit-payloads-testing-branch')
+        uses: actions/checkout@v4
+        with:
+          repository: rapid7/metasploit-payloads
+          path: metasploit-payloads
+          ref: ${{ env.metasploitPayloadsCommit }}
+
+      - name: Build Java & Android payloads
```
Comment (adfoster-r7) [inline .github/workflows/acceptance.yml:247]:
```suggestion
      - name: Build Java and Android payloads
```

File: alans_payload_ci_support_golf.patch
Comment (adfoster-r7) [inline alans_payload_ci_support_golf.patch:1]: 🔥

File: data/cmd_exec/README.md [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  - `show_args_macos`
+
+## Compile binaries locally
+
+We make use of gcc for this: https://hub.docker.com/_/gcc
+
+- Run:
+```shell
+docker run --rm -v "$PWD":/usr/src/myapp -w /usr/src/myapp gcc:11.4.0 /bin/bash -c "apt update && apt install -y gcc-mingw-w64 && make all -f makefile.mk"
+```
```
Comment (adfoster-r7) [inline data/cmd_exec/README.md:31]:
```suggestion

You will need to compile the OSX payload separately on an OSX machine, Docker is not supported
```

File: data/cmd_exec/show_args.c [hunk 14]
```diff
@@ -0,0 +1,8 @@
+int printf(const char *format, ...);
+
+int main(int argc, char *argv[]) {
+
```
Comment (adfoster-r7) [inline data/cmd_exec/show_args.c:4]:
```suggestion
```

**Pull request merged**